core/shared/src/main/scala/cats/effect/unsafe/IORuntimeBuilder.scala

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2020-2021 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cats.effect
+package unsafe
+
+import scala.concurrent.ExecutionContext
+
+/**
+ * Builder object for creating custom `IORuntime`s. Useful for creating `IORuntime` based on the
+ * default one but with some wrappers around execution contexts or custom shutdown hooks.
+ */
+final class IORuntimeBuilder protected (
+    protected var customCompute: Option[(ExecutionContext, () => Unit)] = None,
+    protected var computeWrapper: ExecutionContext => ExecutionContext = identity,
+    protected var customBlocking: Option[(ExecutionContext, () => Unit)] = None,
+    protected var blockingWrapper: ExecutionContext => ExecutionContext = identity,
+    protected var customConfig: Option[IORuntimeConfig] = None,
+    protected var customScheduler: Option[(Scheduler, () => Unit)] = None,
+    protected var extraShutdownHooks: List[() => Unit] = Nil
+) extends IORuntimeBuilderPlatform {
+
+  /**
+   * Override the defaul computing execution context
+   *
+   * @param compute
+   *   execution context for compute
+   * @param shutdown
+   *   method called upon compute context shutdown
+   */
+  def overrideCompute(compute: ExecutionContext, shutdown: () => Unit = () => ()) = {
```

Comment:
```suggestion
  def setCompute(compute: ExecutionContext, shutdown: () => Unit) = {
```

Or, `withCompute`? Not sure what's more idiomatic, but I think `set` is for mutable and `with` is for immutable.

Vasil makes a good point about default parameters. If we really want to make `shutdown` optional we can add an overload. I think we should make it required though, to force users to think about shutdown. It's easy to type `() => ()` anyway.

Reply:
Oh, another thing to consider: what if they `setCompute` twice, should we `shutdown` the previous one? We can either not care (it's a strange thing to do) or we can only allow them to call `setCompute` at most once (if there's already a custom compute set, raise an error).

Reply:
I've renamed all the methods that do override to use the `set` prefix. Allowing to set compute/blocking/shutdown and build just once sounds great. Just to make sure I got you right, do you think we should throw some runtime exception?
Sounds most straightforward, but I wonder if it's worth it to prevent on type-level somehow? or perhaps it would just be an overengineering?

Reply:
Typelevel stuff is cool but probably too over-engineered for this 😆 I was just thinking a runtime exception.

Reply:
Done :wink: